test-data/unit/check-generics.test

```diff
@@ -1815,3 +1815,214 @@ def g(x: T) -> T: return x
+reveal_type(A[str].f())  # E: Revealed type is 'typing.Iterable[Tuple[builtins.str, builtins.str]]'
+[builtins fixtures/classmethod.pyi]
+
+[case testClassMethodOfGenericClassComplexArgumentType]
+from typing import Generic, TypeVar, Iterable, Tuple
+T = TypeVar('T')
+class A(Generic[T]):
+    @classmethod
+    def f(cls, arg: Iterable[Tuple[T, T]]) -> T: pass
+reveal_type(A[str].f((('x', 'y'),)))  # E: Revealed type is 'builtins.str'
+[builtins fixtures/classmethod.pyi]
+
+[case testClassMethodOfGenericClassMoreComplexTypeAndReturnTypeAndArgumentType]
+from typing import Generic, TypeVar, Iterable, Tuple
+T = TypeVar('T')
+class A(Generic[T]):
+    @classmethod
+    def f(cls, arg: Iterable[Tuple[T, T]]) -> Iterable[Tuple[T, T, T]]: pass
+arg = ((('x',), ('y',),),)
+reveal_type(A[Iterable[str]].f(arg))  # E: Revealed type is 'typing.Iterable[Tuple[typing.Iterable[builtins.str], typing.Iterable[builtins.str], typing.Iterable[builtins.str]]]'
+[builtins fixtures/classmethod.pyi]
+
+[case testClassMethodOfGenericClassMultipleTypeVariable]
+from typing import Generic, TypeVar, Tuple
+T = TypeVar('T')
+S = TypeVar('S')
+class A(Generic[T, S]):
+    @classmethod
+    def f(cls) -> Tuple[T, S]: pass
+reveal_type(A[str, int].f())  # E: Revealed type is 'Tuple[builtins.str, builtins.int]'
+reveal_type(A[int, str].f())  # E: Revealed type is 'Tuple[builtins.int, builtins.str]'
+[builtins fixtures/classmethod.pyi]
+
+[case testClassMethodOverloadingGenericClass]
+from typing import overload, TypeVar, Generic, Tuple
+T = TypeVar('T')
+class A(Generic[T]):
+    @overload
+    @classmethod
+    def f(cls, s: str) -> Tuple[str, T]: pass
+    @overload
+    @classmethod
+    def f(cls, s: int) -> Tuple[T, int]: pass
+    @classmethod
+    def f(cls, s): pass
+class B(A[str]): pass
+class C(A[int]): pass
+reveal_type(A[str].f(1))  # E: Revealed type is 'Tuple[builtins.str, builtins.int]'
+reveal_type(A[str].f('a'))  # E: Revealed type is 'Tuple[builtins.str, builtins.str]'
+reveal_type(A[int].f(1))  # E: Revealed type is 'Tuple[builtins.int, builtins.int]'
+reveal_type(A[int].f('a'))  # E: Revealed type is 'Tuple[builtins.str, builtins.int]'
+reveal_type(B.f(1))  # E: Revealed type is 'Tuple[builtins.str*, builtins.int]'
+reveal_type(B.f('a'))  # E: Revealed type is 'Tuple[builtins.str, builtins.str*]'
+reveal_type(C.f(1))  # E: Revealed type is 'Tuple[builtins.int*, builtins.int]'
+reveal_type(C.f('a'))  # E: Revealed type is 'Tuple[builtins.str, builtins.int*]'
+[builtins fixtures/classmethod.pyi]
+
+[case testClassmethodFromClassMethodOfGenericClass]
+from typing import Generic, TypeVar
+T = TypeVar('T')
+class X(Generic[T]):
+    @classmethod
+    def a(cls) -> T: pass
+    @classmethod
+    def b(cls) -> T:
+        return cls.a()
+X.a  # E: Access to classmethod of generic type remains type variables T is ambiguous.
+reveal_type(X[str].b())  # E: Revealed type is 'builtins.str'
+[builtins fixtures/classmethod.pyi]
```

Comment:
Thanks for adding many tests! A little note on tests, they will be easier to read if you add few (but only a few) newlines in each separating some logical blocks.

Reply:
Thank you. I didn't have the confidence. Because I feel that current code bases are using them only fewer.

I will add.